### ISSUE TITLE
dnf: UnicodeDecodeError when running `make build'

### ISSUE THREAD

**Issue by rakuco** [clearlinux/distribution#56]
I'm on version 22180, and `dnf` is failing whenever I run `make build` with a Clear Linux package.

For example, `cd`'ing to `packages/tmux` and running `make build' results in the following error after installing the package's build dependencies:

```
Traceback (most recent call last):
  File "/usr/bin/dnf", line 58, in <module>
    main.user_main(sys.argv[1:], exit_code=True)
  File "/usr/lib/python3.6/site-packages/dnf/cli/main.py", line 179, in user_main
    errcode = main(args)
  File "/usr/lib/python3.6/site-packages/dnf/cli/main.py", line 64, in main
    return _main(base, args, cli_class, option_parser_class)
  File "/usr/lib/python3.6/site-packages/dnf/cli/main.py", line 99, in _main
    return cli_run(cli, base)
  File "/usr/lib/python3.6/site-packages/dnf/cli/main.py", line 123, in cli_run
    ret = resolving(cli, base)
  File "/usr/lib/python3.6/site-packages/dnf/cli/main.py", line 154, in resolving
    base.do_transaction(display=displays)
  File "/usr/lib/python3.6/site-packages/dnf/cli/cli.py", line 238, in do_transaction
    super(BaseCli, self).do_transaction(display)
  File "/usr/lib/python3.6/site-packages/dnf/base.py", line 784, in do_transaction
    self._plugins.run_transaction()
  File "/usr/lib/python3.6/site-packages/dnf/plugin.py", line 96, in fn
    dnf.util.mapall(operator.methodcaller(method), self.plugins)
  File "/usr/lib/python3.6/site-packages/dnf/util.py", line 223, in mapall
    return list(map(fn, *seq))
  File "/usr/lib/python3.6/site-packages/dnf-plugins/show_leaves.py", line 51, in transaction
    for x in self.leaves_command.findleaves()))
  File "/usr/lib/python3.6/site-packages/dnf-plugins/leaves.py", line 150, in findleaves
    (packages, depends, rdepends) = self.buildgraph()
  File "/usr/lib/python3.6/site-packages/dnf-plugins/leaves.py", line 55, in buildgraph
    sreq = str(req)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x8a in position 3: invalid start byte
```

**Comment by Gabib32** [comment]
@rakuco I actually ran into the same thing yesterday. Try running a 'make mockclean' and 'make mockproper' in the top level clear/ directory to clean up the cache

**Comment by rakuco** [comment]
Thanks, `make mockproper` did the trick for me. I'm leaving this open in case the Clear team wishes to address the underlying bug here.

**Comment by iphutch** [comment]
@phmccarty is this something the tooling can auto-detect and fix?

**Comment by phmccarty** [comment]
I will begin looking into this issue

**Comment by phmccarty** [comment]
The issue appears to be related to DNF's handling of the root cache when there are newer RPMs available to "update" the ones extracted from the cache. This explains why a `make mockproper` is a solid workaround, because it removes the root cache (/var/cache/mock/clear/root_cache/cache.tar.gz).

Still investigating.

**Comment by phmccarty** [comment]
I have not tracked down the root cause yet. Running `make mockproper` is still a solid workaround, so this can be considered lower priority.

**Comment by phmccarty** [comment]
I spent more time debugging this issue last week, and my current suspicion is that libdnf is to blame, since the value failing to be cast to a UTF-8 encoded string originates from hawkey, which is part of libdnf.

Much of the libdnf data originates from libsolv though. I am still slightly suspicious of libsolv, since it's a dependency of libdnf, and it's possible that libdnf could be passing along corrupt data. But last week, I ran `dumpsolv` on /var/cache/mock/dnf_cache/*.solv, and I didn't notice any corruption in the dependency string(s) that led to the exception. So I am going to ignore the libsolv scenario for now. And the exception is still raised with the latest libsolv (0.6.34), so no change of behavior there.

The exception is still raised with latest mock (1.4.11), though because mock wraps a bunch of other tools that are doing most of the work, I doubt mock is to blame.

I will be testing latest dnf and libdnf next to see if behavior is different. Other dependencies include librepo and the rpm python bindings, but I'm doubtful those would be to blame.

(This debug effort is slow-going, because the bug is very intermittent... Running `make mockclean` always makes the issue go away for me, and `make MOCK_OPTS='--offline'` sometimes fixes the issue, but only temporarily.)

**Comment by rakuco** [comment]
Thank you for taking care of this!

**Comment by phmccarty** [comment]
I wanted to post a followup on this issue for those that are still monitoring it:

TL;DR - The libsolv package was configured in a way that may lead to corruption of the tracking information for the local rpmdb, which in turn may have resulted in the UnicodeDecodeError. I am updating libsolv to more closely match Fedora's packaging, which makes libsolv use librpm instead of libdb to parse the local rpmdb. In my testing of this change, I am seeing no longer seeing the corruption.

Details:

Upgrading DNF back in July did not actually resolve the issue... The reason why I was not seeing it was because my local mock root cache was being regenerated all the time, thus containing the very latest set of packages, but the exception only appears for a root cache containing packages that must be upgraded.

I spent a lot of time last week debugging the issue looking for the root cause, tracing through dnf/libdnf/libsolv. I discovered that the rpmdb data, as tracked by DNF, may become corrupt, leading to the UnicodeDecodeError if the corruption yields invalid UTF-8 sequences within any package's "requires" dependency data. Then after much digging, I discovered that when the corruption appeared, it appeared immediately after the RPM transaction ("update" in this case) runs within the build root.

Tracing through the libdnf and libsolv code to see how the local rpmdb is read from disk, I found that libsolv offers two alternatives for reading it, using either libdb or librpm. Reading with libdb is the default, and reading with librpm is an alternative that can be configured with a build-time cmake option. 

After inspecting the Fedora and OpenSUSE packaging of libsolv, I noticed that Fedora sets the librpm option, but OpenSUSE does not. Thinking that Clear Linux usage of libsolv should be more closely aligned to Fedora because we use DNF, I suspected that we should configure libsolv to use librpm instead. So I applied the change the libsolv, rebuilt libdnf against it, tested it, and I saw the corruption (and also the UnicodeDecodeError) disappear.

This still feels like a workaround, since libsolv appears to be (sometimes) misreading the rpmdb when using libdb to do so. But at the same time, this the code path used by OpenSUSE's libsolv, so I would expect that code to be known-working. Unless of course there are others factors at play, like needing to configure libdb differently. Regardless, the librpm code path for libsolv appears to be working well on Clear Linux, so I will settle on that for now.

**Comment by rakuco** [comment]
This is awesome; I ran into this issue about a month ago but thought it was a one-off error at the time. Again, thanks a lot for digging so deep for the root cause here.
